### PR TITLE
Add socket connection and example TCP usage

### PR DESCRIPTION
**Pull request by sectioneight**
This is subject to change, but it shows how we would exchange and encode/decode messages over a `socket`

@uber/soap

### CODE REVIEW

File: tchannel/frame.py [hunk 1]
```diff
@@ -1,21 +1,52 @@
...
+        message_id = read_number(stream, cls.ID_WIDTH)
+        message_type = read_number(stream, cls.TYPE_WIDTH)
+        message_class = get_message(message_type)
+        if not message_class:
+            raise ProtocolException('Unknown message type: %d' % message_type)
+
+        flags = read_number(stream, cls.FLAGS_WIDTH)
+        if flags:
+            # TODO handle partial messages (e.g. multiple frames)
+            pass
```
Comment (abhinav) [inline tchannel/frame.py]: Use `raise NotImplementedError()` instead.

File: tchannel/mapping.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    PingResponseMessage,
+]
+
+MESSAGE_TYPES_TO_CLASSES = {
+    msg.message_type: msg
+    for msg in ALL_MESSAGES
+}
+
+
+def get_message(message_type):
```
Comment (abhinav) [inline tchannel/mapping.py]: This should probably be named `get_message_class`. The current name suggests that it gets an actual message.

File: tchannel/frame.py [hunk 3]
```diff
@@ -1,21 +1,52 @@
...
     ID_WIDTH = 4
     TYPE_WIDTH = 1
+    FLAGS_WIDTH = 1
     PRELUDE_SIZE = 0x10  # this many bytes of framing before payload
     RESERVED_PADDING = b'\x00\x00\x00\x00\x00\x00'  # 6 bytes are reserved
+    RESERVED_WIDTH = len(RESERVED_PADDING)
 
     def __init__(self, message, message_id):
         self._message = message
         self._message_id = message_id
```
Comment (abhinav) [inline tchannel/frame.py]: Do these need to be protected with the `_`? Are there no cases where we'd want to access the message from the `Frame` object?

File: tchannel/frame.py (hunk 1 above)
Comment (abhinav) [inline tchannel/frame.py]: Also, my understanding is that there are only a few message types that will allow having the flag set?

Comment (sectioneight) [inline tchannel/frame.py]: The flag is a single-byte, which means there are 8 bits of flagging that could be used. Currently we have a single flag (the low bit) that means there are more frames coming for this message ID.

File: tchannel/mapping.py (hunk 2 above)
Comment (sectioneight) [inline tchannel/mapping.py]: Definitely. Will update.

File: tchannel/frame.py (hunk 3 above)
Comment (sectioneight) [inline tchannel/frame.py]: I think I want to actually hide the fact that we're using framing at all, and return a "context" to the people who speak the protocol, where the context will have the message ID, the calling service name etc. So yea, I don't expect people to be pulling the message_id off the frame explicitly.

File: tests/test_socket.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import contextlib
+import socket
+
+import pytest
+
+from tchannel.socket import Connection
+
+
+@pytest.fixture
+@contextlib.contextmanager
```
Comment (abhinav) [inline tests/test_socket.py]: Any reason not to use `pytest.yield_fixture` instead? Takes away the need to use the context manager in the test.

Comment (sectioneight) [inline tests/test_socket.py]: Never heard of that. You are so full of awesome tips. Thanks!

**Comment by abhinav** [comment]
:shipit:

**Pull request merged**